sleigh/sleigh-compile/src/matcher/mod.rs

```diff
@@ -19,421 +21,304 @@ pub(crate) fn build_sequential_matcher(
     Ok(SequentialMatcher { cases, token_size })
 }
 
-enum Index {
-    Token(u8),
-    Context(u8),
+struct IndexLinkedList<'a, T> {
+    head: Option<usize>,
+    order: Vec<Option<usize>>,
+    elements: &'a [T],
 }
 
-struct BitConstraint {
-    index: Index,
-    ones: u32,
-    zeroes: u32,
-    unconstrained: u32,
+impl<'a, T> IndexLinkedList<'a, T> {
+    fn new_empty(elements: &'a [T]) -> Self {
+        let order = Vec::with_capacity(elements.len());
+        Self { head: None, order, elements }
+    }
+    fn final_order(self) -> impl Iterator<Item = usize> {
+        struct Iter(Vec<Option<usize>>, Option<usize>);
+        impl Iterator for Iter {
+            type Item = usize;
+            fn next(&mut self) -> Option<Self::Item> {
+                let next = self.0[self.1?];
+                if let Some(next) = next {
+                    self.1 = Some(next);
+                }
+                next
+            }
+        }
+        Iter(self.order, self.head)
+    }
+
+    fn insert_at(&mut self, pred: impl Fn(&'a T) -> bool) {
```

Comment:
Maybe rename to `insert_before`. Since I think this inserts the element before the first index where `pred` is true.

Reply:
Fixed